src/test/common/test_fair_mutex.cc

```diff
@@ -61,11 +61,8 @@ TEST(FairMutex, fair)
       };
     }
   };
-  std::array<std::thread, NR_TEAMS> teams;
+  std::array<std::future<void>, NR_TEAMS> completed;
   for (int team = 0; team < NR_TEAMS; team++) {
-    teams[team] = std::thread(play, team);
-  }
-  for (auto& team : teams) {
-    team.join();
+    completed[team] = std::async(std::launch::async, play, team);
```

Comment:
Isn't there no guarantee that these actually execute asynchronously (i.e. no threads may be spawned)?

Reply:
@batrick i was referencing [cppreference.com](https://en.cppreference.com/w/cpp/thread/async):

> If the async flag is set (i.e. (policy & std::launch::async) != 0), then async executes the callable object f on a new thread of execution (with all thread-locals initialized) as if spawned by std::thread(std::forward<F>(f), std::forward<Args>(args)...), except that if the function f returns a value or throws an exception, it is stored in the shared state accessible through the std::future that async returns to the caller.

in which *a new thread* is spawned for executing the callable. so i think it's guaranteed.

Reply:
I see, this is new in C++17 apparently? IIRC from a C++ book, `std::async` was almost DOA because you had no guarantee threads would actually be used.

Anyway, LGTM.

Reply:
not sure. the policy parameter was introduced by C++11. but cppreference does not claim that its behavior changed in C++17. anyway, since we are using GCC-8 (el8) and GCC-10 (ubuntu focal). guess it's safe it is well supported in GCC-8 and up per https://gcc.gnu.org/projects/cxx-status.html#cxx17.